lib/curl_sha512_256.c

```diff
@@ -0,0 +1,596 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Evgeny Grin (Karlson2k), <k2k@narod.ru>.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if !defined(CURL_DISABLE_DIGEST_AUTH) && !defined(CURL_DISABLE_SHA512_256)
+
+#include "curl_sha512_256.h"
+#include "warnless.h"
+
+/* ** This implementation of SHA-512/256 hash calculation was originally ** *
+ * ** written by Evgeny Grin (Karlson2k) for GNU libmicrohttpd.          ** *
+ * ** The author ported the code to libcurl. The ported code is provided ** *
+ * ** under curl license.                                                ** *
+ * ** This is a minimal version with minimal optimisations. Performance  ** *
+ * ** can be significantly improved. Big-endian store and load macros    ** *
+ * ** are obvious targets for optimisation.                              ** *
+ * ** The code is designed in a way so any wide enough type (64 bits or  ** *
+ * ** more) can be used for calculations. Potentially it can be larger   ** *
+ * ** then 64 bits.                                                      ** */
+
+#ifdef SIZEOF_INT
+#if SIZEOF_INT == 8
+#define CURLX_UINT64 unsigned int
+#define CURLX_UINT64_C(val) val##U
+#endif /* SIZEOF_INT == 8 */
+#endif /* SIZEOF_INT */
```

Comment:
Not at the moment.
The code needs to be future-proof. 
The same for long long.

Reply:
you have not tested on any such platform have you, so how do you *know* it is future-proof? You're only adding code that is never reached/used. Lots of code everywhere else in curl is going to break miserably long before his code is reached if anyone would ever introduce such weird stuff.

Reply:
> `int` is never 8 bytes on any platform

Not correct.
ILP64 data model already exists.
See https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

Reply:
Read again. There are no platforms with `int` larger than 32 bits in current use where anyone would build curl. It's the same kind of argument saying that `char` does not have to be 8 bits etc. They are just theoretical proofs that it is *possible*, but they are not used.

Since the code is not tested on any such machines anyway, claiming that it works on such machines is just wishful thinking.

Reply:
OK, I'll remove it. No problem.

Reply:
Just wanted to help you have one less place to fix when it happens. :)